bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/GarbageCollectorThread.java

```diff
@@ -335,7 +246,10 @@ public void safeRun() {
         if (force) {
             LOG.info("Garbage collector thread forced to perform GC before expiry of wait time.");
         }
-
+        // Recover and clean up previous state if using transactional compaction
+        if (compactor instanceof TransactionalEntryLogCompactor) {
+            ((TransactionalEntryLogCompactor) compactor).cleanUpAndRecover();
```

Comment:
It is better to add an empty default method _cleanUpAndRecover_ to AbstractLogCompactor and override it in TransactionalEntryLogCompactor .  It would be a more "object oriented"

Reply:
That's a good idea, fixed.